tuning_test.py

```diff
@@ -0,0 +1,266 @@
+# coding=utf-8
+# Copyright 2021 The Google Research Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Tests for synthetic_protein_landscapes.tuning."""
+
+from absl.testing import absltest
+from absl.testing import parameterized
+
+import numpy as np
+from scipy.special import comb
+
+import tuning
+import potts_model
+import sampling
+
+
+class TuningParamsTest(parameterized.TestCase):
+    """Test class for tuning."""
+
+    def _get_params(self, seed):
+        """Weight matrix and field vector."""
+        rng = np.random.default_rng(seed)
+        weight_matrix = rng.standard_normal(size=(4, 4, 20, 20), dtype=np.float32)
+        # make symmetric
+        weight_matrix = weight_matrix + np.moveaxis(weight_matrix, (0, 1, 2, 3), (1, 0, 3, 2))
+        field_vec = rng.standard_normal(size=(4, 20), dtype=np.float32)
+        return weight_matrix, field_vec
+
+    def _get_landscape(self, seed, wt_seq=[0, 0, 0, 0], **kwargs):
+        """Return a small PottsModel landscape."""
+        weight_matrix, field_vec = self._get_params(seed)
+        return potts_model.PottsModel(weight_matrix, field_vec, wt_seq=wt_seq, **kwargs)
+
+    @parameterized.named_parameters(
+        dict(
+            testcase_name='wt_0',
+            seed=0,
+            wt_seq=[0, 0, 0, 0],
+        ),
+        dict(
+            testcase_name='wt_1',
+            seed=1,
+            wt_seq=[1, 1, 1, 1],
+        ),
+    )
+    def test_normalize_to_singles(self, wt_seq, seed):
+        untuned_landscape = self._get_landscape(wt_seq=wt_seq, seed=seed)
+        tuning_kwargs = tuning.get_tuning_kwargs(untuned_landscape, normalize_to_singles=True)
+        tuned_landscape = self._get_landscape(wt_seq=wt_seq, seed=seed, **tuning_kwargs)
+
+        all_single_fitness = tuned_landscape.evaluate(
+            sampling.get_all_single_mutants(wt_seq, tuned_landscape.vocab_size))
+        np.testing.assert_allclose(np.std(all_single_fitness), 1.0, rtol=1e-6)
```

Comment:
Just a note: you may have to mess around with that tolerance if you're running on TPU with 16 bits of precision.

Reply:
Yeah, I wish this were simpler to maintain robustness to floating point precision. I guess for now I'll assume np.float32 everywhere.

I think the likelihood that we end up on TPUs is pretty slim, but I appreciate the vote of confidence :)